### DISCUSSION TITLE
Apple OAuth2 on Dart SDK doesn't return user's email

### DISCUSSION THREAD

**Discussion by tonnax** [pocketbase/pocketbase#3814] [Q&A]
Hi dev,

I've been debugging this issue for several days now and still can't figure it out how to 'required' Sign In with Apple to return user's Apple ID email for my app. Sign In with Google is ok, they give me all the necessary things for my user registration datas including the user's email.

I'm using Dart SDK for my Flutter mobile app. Already tried using SignInWithApple package and still not sure how to sync it with the SDK. Already try to digest this https://github.com/pocketbase/pocketbase/discussions/1591 but can't relate it with the Dart SDK. 

Thanks in advance team!

**Comment by CogalTek** [comment]
Hi, I have a hypothesis, maybe you need to have an Apple developer license to be able to retrieve all the information you need, I know that for Microsoft you need a business account, so it wouldn't surprise me that this is the case at apple too

**Reply by tonnax** [reply]
Hi @CogalTek thanks for your help. 

I already have Apple developer account and already successfully integrate Apple OAuth2 into my app, but it didn't return users email. I was wondering how fellow mobile devs here that are using Pocketbase overcome this issue.

**Reply by tonnax** [reply]
`final authData =
                            await pb.collection('users').authWithOAuth2(
                          'apple',
                          (url) async {
                            var newUrl = Uri.parse(
                                url.toString() + "&response_mode=form_post");
                            if (await canLaunchUrl(newUrl)) {
                              print(newUrl);
                              print(url);
                              await launchUrl(newUrl);
                            } else {
                              throw 'Could not launch $newUrl';
                            }
                          },
                          scopes: const ['email'],
                        );`

 I tried to add the scope and response_mode, but now it return 404 after initially asking my Apple ID email, which is not the case before adding the email scope.

**Reply by ganigeorgiev** [reply]
@tonnax You usually shouldn't need to specify explicitly the "email" scope, bit if you do you'll need also `openid`.

See also the note from the [Apple docs](https://developer.apple.com/documentation/sign_in_with_apple/sign_in_with_apple_rest_api/authenticating_users_with_sign_in_with_apple):

> The email address is either the user’s real email address or the proxy address, depending on their private email relay service. This value may be empty for Sign in with Apple at Work & School users. For example, younger students may not have an email address.

It could be also some configuration option in the Apple developer dashboard OR on the device itself, but unfortunately I don't have access to iOS device or Apple developer account to test if this is the case.

Also please note that the Apple OAuth2 integration was based entirely on whatever I could find in their docs without actually testing it since I don't have Apple developer account, so it is also possible that I've missed something (see PR https://github.com/pocketbase/pocketbase/pull/1954).

**Reply by tonnax** [reply]
<img width="1413" alt="image" src="https://github.com/pocketbase/pocketbase/assets/140943390/e8710e51-c9f9-4d2a-a8e8-d19346b07340">

This is the details on the code and response I get on 3 of my approach. [@n1xx1](https://github.com/n1xx1) mention this before back then in a [discussion](https://github.com/pocketbase/pocketbase/issues/202#issuecomment-1300435689).

@ganigeorgiev I acknowledge that sometime user might register using Apple's masked email, but even so, that masked email will be store in DB as well. Now the value I got is always N/A. I have the Apple Dev account so let me know if you got time, I can help you with the testing! 💪

**Reply by ganigeorgiev** [reply]
@tonnax I don't think I'll be able to help you, sorry.

The Apple integration already diverge a lot from the other OAuth2 providers and to be honest I'm not sure that I like the idea to add further special handling just because it is Apple, but if the required changes are not so drastic I'll accept a PR.

If `response_mode=form_post` is really required for returning the email (_based on my understanding of the Apple docs it shouldn't be needed if no explicit scopes are defined_) and if someone wants to take a look at this:

- Update the default Apple auth url to use `response_mode=form_post` in https://github.com/pocketbase/pocketbase/blob/985ab1e5b79a328b516decca3d1d7dc8edfcbba0/apis/record_auth.go#L140 (this can be changed also client-side before launching the Apple auth url).

- Add support for both GET and POST in the default redirect handler in https://github.com/pocketbase/pocketbase/blob/985ab1e5b79a328b516decca3d1d7dc8edfcbba0/apis/record_auth.go#L31 (need to be verified that in the case of POST, the `code` and `state` parameters are still send as a query params, if not - the related action will also will have to be updated to accommodate this. For those who use the "Manual code exchange" flow and have their own redirect handler they will have to do the same.

- If changes on the provider/user data fetching is needed, the Apple provider code can be found in https://github.com/pocketbase/pocketbase/blob/master/tools/auth/apple.go.

**Reply by tonnax** [reply]
Thanks @ganigeorgiev I hope we can have a ticket for this before v1 (or maybe one day) since Apple is "[forcing](https://appleinsider.com/articles/19/06/03/sign-in-with-apple-mandatory-for-all-apps-with-third-party-sign-in-options)" all iOS apps to use Sign-In with Apple. 

No offense, but Pocketbase selling point is that it will be the next great all-in-one backend for **mobile app** and Apple iOS is a leading OS in smartphone industry. I really hope as a mobile developer myself, Pocketbase can make open source mobile development easier than ever.

I'm just learning to code, really want to help, but tweaking Pocketbase is like a God-level programming for me, not yet there. Peace. Really grateful for your invention man, keep up the good work! 💪 🎊

**Reply by ganigeorgiev** [reply]
_Side-note: I've pushed a small change in v0.22.14-rc that adds support for `response_mode=form_post` but needs help with the testing. If someone have time please check https://github.com/pocketbase/pocketbase/discussions/5074._
